src/main/java/com/google/devtools/common/options/Converters.java

```diff
@@ -625,4 +632,71 @@ public String getTypeDescription() {
       return "Converts to a CacheBuilderSpec, or null if the input is empty";
     }
   }
+
+  public static class BooleanFormulaConverter implements Converter<Expression> {
```

Comment:
(I don't know why GitHub's code review tool won't let me continue the previous discussion. Is it because the relevant declaration moved? In any case, not helpful.)

You wrote:
> Making this class private will not work since the methods convert and getTypeDescription must be accessible outside of LoadingOptions. Do you still want it to be moved but stay public or is it better if it stays in Converters?

The class can be private, but its methods must be public because they override public methods defined in a superclass. This is quite normal practice in Java.

Please do move it out of Converters, for the reason I gave previously.

Reply:
I think I have misunderstood what you meant then. 
I thought you wanted me to move the class **BooleanFormulaConverter** to **LoadingOptions** like this:
```
public class LoadingOptions extends OptionsBase {
  ...

  @Option(
    name = "build_tag_filters",
    converter = BooleanFormulaConverter.class,
    defaultValue = "",
    documentationCategory = OptionDocumentationCategory.UNCATEGORIZED,
    effectTags = {OptionEffectTag.UNKNOWN},
    help =
        "See '--test_tag_filters' for instructions. "
            + "This option does not affect the set of tests executed with the 'test' command; "
            + "those are to be governed by the test filtering options, for example '--test_tag_filters'"
  )
  public Expression buildTagFilter;

  @Option(
    name = "test_tag_filters",
    converter = BooleanFormulaConverter.class,
    defaultValue = "",
    documentationCategory = OptionDocumentationCategory.UNCATEGORIZED,
    effectTags = {OptionEffectTag.UNKNOWN},
    help =
        "Specifies a boolean formula that follows the syntax: "
            + "expr = IDENT | expr 'or' expr | expr 'and' expr | 'not' expr | '(' expr ')'. "
            + "Only those targets will be built that contain at least one included tag and do not contain any "
            + "excluded tags. This option affects --build_tests_only behavior and the test command."
            + "Example of boolean formula: tag1 and (tag2 or not tag3). "
            + "NOTE: To specify a list of tags separated by comma is legacy syntax and will not be supported "
            + "after Jan 1, 2021."
  )
  public Expression testTagFilter;

  ...

  private static class BooleanFormulaConverter implements Converter<Expression> {
    @Override
    public Expression convert(String input) throws OptionsParsingException {
      if (input.isEmpty()) {
        return Expression.empty();
      }
      if (isLegacySyntax(input)) {
        input = convertLegacySyntaxToBooleanFormula(input);
      }
      try {
        Expression result = Expression.parse(ParserInput.fromLines(input));
        validate(result);
        return result;
      } catch (SyntaxError.Exception e) {
        throw new OptionsParsingException("Failed to parse expression: " + e.getMessage() + " input: " + input, e);
      }
    }

    @Override
    public String getTypeDescription() {
      return "A Boolean formula over test tags (e.g. '(release or smoke) and not flaky')";
    }

    private static boolean isLegacySyntax(String input) {
      return input.contains(",") || input.contains("-");
    }

    /**
     * Converts a string containing tags separated by comma to a boolean expression.
     */
    private String convertLegacySyntaxToBooleanFormula(String input) throws OptionsParsingException {
      return input.replaceAll(" *, *", " or ").replace("-", " not ").trim();
    }

    /**
     * Throws an OptionsParsingException when the given boolean formula does not follow the grammar:
     * expr = IDENT | expr 'or' expr | expr 'and' expr | 'not' expr | '(' expr ')'
     */
    private void validate(Expression expression) throws OptionsParsingException {
      switch (expression.kind()) {
        case IDENTIFIER:
          break;
        case BINARY_OPERATOR:
          BinaryOperatorExpression boe = (BinaryOperatorExpression) expression;
          if (boe.getOperator() == TokenKind.OR || boe.getOperator() == TokenKind.AND) {
              validate(boe.getX());
              validate(boe.getY());
          } else {
            throw new OptionsParsingException(String.format("invalid Boolean operator: %s (want 'and' or 'or')",
                boe.getOperator()));
          }
          break;
        case UNARY_OPERATOR:
          UnaryOperatorExpression uoe = (UnaryOperatorExpression) expression;
          if (uoe.getOperator() == TokenKind.NOT) {
            validate(uoe.getX());
          } else {
            throw new OptionsParsingException(String.format("invalid Boolean operator: %s (want 'not')",
                uoe.getOperator()));
          }
          break;
        default:
          throw new OptionsParsingException("invalid Boolean operator");
      }
    }
  }
}
```

This leads to a build failure which does not occur when the visibility of **BooleanFormulaConverter** is set to public.

The error message is:
```
ERROR: /repo/ehkloaj/bbi_bazel/bazel/src/main/java/com/google/devtools/build/lib/BUILD:498:1: Executing genrule //src/main/java/com/google/devtools/build/lib:gen_command-line-reference failed (Exit 37): bash failed: error executing command 
  (cd /repo/ehkloaj/bazel_build_root/584b087bddb32230268955ea080f433e/sandbox/processwrapper-sandbox/3904/execroot/io_bazel && \
  exec env - \
    PATH=/app/vbuild/RHEL7-x86_64/python/3.8.0/bin:/app/vbuild/RHEL6-x86_64/jdk/1.8.0_162-1/bin:/opt/tools/tools/vulcanbase/0.74/bin:/app/vbuild/SLED11-x86_64/openjdk/12.0.2/bin:/app/vbuild/RHEL7-x86_64/idea/2019.3.3/bin:/app/vbuild/RHEL7-x86_64/git/2.9.3-1/bin:/app/vbuild/RHEL7-x86_64/git/2.9.3-1/libexec/git-core:/env/sero/bin:/home/ehkloaj/.afs/0/ibin:/usr/afsws/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/sbin:/sbin:/app/arc/0/bin \
    python3_dir=/app/vbuild/RHEL7-x86_64/python/3.8.0/bin \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; cat site/command-line-reference-prefix.html > bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html && TMP=`mktemp -d /tmp/tmp.XXXXXXXXXX` && export JARBIN=external/remotejdk11_linux/bin/jar && bazel-out/host/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer --jvm_flag=-Dio.bazel.EnableJni=0 --batch --install_base=${TMP} --output_base=${TMP}/output/ --output_user_root=${TMP} --failure_detail_out=${TMP}/output/failure_detail.rawproto help everything-as-html >> bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html 2>/dev/null && cat site/command-line-reference-suffix.html >> bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html')
Execution platform: //:default_host_platform

Use --sandbox_debug to see verbose messages from the sandbox bash failed: error executing command 
  (cd /repo/ehkloaj/bazel_build_root/584b087bddb32230268955ea080f433e/sandbox/processwrapper-sandbox/3904/execroot/io_bazel && \
  exec env - \
    PATH=/app/vbuild/RHEL7-x86_64/python/3.8.0/bin:/app/vbuild/RHEL6-x86_64/jdk/1.8.0_162-1/bin:/opt/tools/tools/vulcanbase/0.74/bin:/app/vbuild/SLED11-x86_64/openjdk/12.0.2/bin:/app/vbuild/RHEL7-x86_64/idea/2019.3.3/bin:/app/vbuild/RHEL7-x86_64/git/2.9.3-1/bin:/app/vbuild/RHEL7-x86_64/git/2.9.3-1/libexec/git-core:/env/sero/bin:/home/ehkloaj/.afs/0/ibin:/usr/afsws/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/sbin:/sbin:/app/arc/0/bin \
    python3_dir=/app/vbuild/RHEL7-x86_64/python/3.8.0/bin \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; cat site/command-line-reference-prefix.html > bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html && TMP=`mktemp -d /tmp/tmp.XXXXXXXXXX` && export JARBIN=external/remotejdk11_linux/bin/jar && bazel-out/host/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer --jvm_flag=-Dio.bazel.EnableJni=0 --batch --install_base=${TMP} --output_base=${TMP}/output/ --output_user_root=${TMP} --failure_detail_out=${TMP}/output/failure_detail.rawproto help everything-as-html >> bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html 2>/dev/null && cat site/command-line-reference-suffix.html >> bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/command-line-reference.html')
Execution platform: //:default_host_platform

Use --sandbox_debug to see verbose messages from the sandbox
INFO: Elapsed time: 37.507s, Critical Path: 22.59s
INFO: 478 processes: 106 processwrapper-sandbox, 372 worker.
FAILED: Build did NOT complete successfully
```

I thought that this was because **BooleanFormulaConverter** is not accessible when the command line reference is generated due to its private visibility but I might be mistaken. 

Also the method **convert** will need to be called outside of the class **LoadingOptions** and I think this will fail if **BooleanFormulaConverter** is private or am I mistaken here as well?

Could you please explain what you meant in more detail, thanks.

Reply:
Well, that is a truly terrible error message: it has suppressed the part the really matters. (Does running with --sandbox_debug show you a Java stack trace?)

I tried it locally and it appears that the class must be declared public for it to be accessible to reflection; the "bazel help everything-as-html" command that failed above uses reflection to print all the flags.

So, please keep the class inside LoadingOptions, but declare it public, and precede it by a comment such as this:

/** BooleanFormulaConverter parses a boolean formula to an Expression. Declared public to make it accessible to reflection. */

Reply:
> Well, that is a truly terrible error message: it has suppressed the part the really matters. (Does running with --sandbox_debug show you a Java stack trace?)
No, same message.